src/orange/construct/detail/NodeReplacementInserter.hh

```diff
@@ -0,0 +1,142 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/construct/detail/NodeReplacementInserter.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <variant>
+#include <vector>
+
+#include "../CsgTree.hh"
+#include "../CsgTypes.hh"
+
+namespace celeritas
+{
+namespace csg
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Add a node ID and its "replaced" value.
+ *
+ * This implementation detail of the "replace down" algorithm adds daughters
+ * to a queue of nodes to visit, along with their replacement values.
+ */
+class NodeReplacementInserter
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using VecNode = std::vector<std::pair<NodeId, Node>>;
+    //!@}
+
+  public:
+    // Construct with references to the stack and the replacement value
+    inline NodeReplacementInserter(VecNode* stack, Node const& repl);
+
+    // If replacing a queued node with a boolean, ours should match
```

Comment:
This comment doesn't make sense without the context that the function itself has a DBC check. I recommend adding standard doc strings to these functions.